File: 코드/ch6/clip5/boj9663/main.cpp
#include <iostream>

int N;
int queenPosition[14];

bool isValid(int y, int x)
{
    for(int i = 0; i < y; i++)
    {
        if(queenPosition[i] == x || y - i == abs(x - queenPosition[i]))
        {
            return false;
        }
    }

    return true;
}

int func(int y) {
    int count = 0;

	if (y == N)
    {
		count = 1;
    }
    else
    {
        for(int x = 0; x < N; x++)
        {
            if(isValid(y, x) == true)
            {
                queenPosition[y] = x;
                count += func(y + 1);
            }
        }
    }
    
	return count;
}

int main() {
	scanf("%d", &N);
	printf("%d", func(0));
	return 0;
}

File: 코드/ch6/clip5/boj9663/Main.java
import java.util.Scanner;

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        System.out.println(solve(n, 0));
    }

    static int[] queen = new int[15];

    static int solve(int n, int row) {
        int count = 0;
        //base case
        if (row == n) return 1;

        // recursive case
        for (int col = 0; col < n; col++) {
            if (isValid(row, col)) {
                queen[row] = col;
                count += solve(n, row + 1);
            }
        }
        return count;
    }

    static boolean isValid(int row, int col) {
        for (int i = 0; i < row; i++) {
            if (queen[i] == col) return false;
            if (Math.abs(row - i) == Math.abs(col - queen[i])) return false;
        }
        return true;
    }
}